# A. Web Scraping

In [87]:
'''
=================================================
Graded Challenge 3

Nama  : Irfan Risqy
Batch : BSD - 001

Program ini dibuat untuk melakukan analisa data menggunakan central tendency, std deviasi, uji Skewness dan Kurtosis, mencari lower dan upper, serta cek korelasi dari data hasil web scraping situs Tokopedia
=================================================
'''

'\n=================================================\nGraded Challenge 3\n\nNama  : Irfan Risqy\nBatch : BSD - 001\n\nProgram ini dibuat untuk melakukan analisa data menggunakan central tendency, std deviasi, uji Skewness dan Kurtosis, mencari lower dan upper, serta cek korelasi dari data hasil web scraping situs Tokopedia\n=================================================\n'

In [20]:
# Setup Library
import requests
import pandas as pd
import time
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.common.by import By


In [21]:

driver = webdriver.Chrome()

url="https://www.tokopedia.com/search?navsource=&page=1&q=seblak&srp_component_id=02.01.00.00&srp_page_id=&srp_page_title=&st="
driver.get(url)
html = driver.page_source

soup = BeautifulSoup(driver.page_source, "html.parser")
print(soup.prettify()[:700])

<html data-rh="lang,translate" lang="id" translate="no">
 <head>
  <title>
   Jual seblak | Tokopedia
  </title>
  <meta charset="utf-8"/>
  <meta content="initial-scale=1, minimum-scale=1, maximum-scale=5, user-scalable=yes, width=device-width" name="viewport"/>
  <link crossorigin="" href="https://assets.tokopedia.net" rel="preconnect"/>
  <link href="https://assets.tokopedia.net" nonce="" rel="dns-prefetch"/>
  <link crossorigin="" href="https://gql.tokopedia.com" rel="preconnect"/>
  <link href="https://gql.tokopedia.com" nonce="" rel="dns-prefetch"/>
  <link crossorigin="" href="https://googleads.g.doubleclick.net" rel="preconnect"/>
  <link href="https://googleads.g.doubleclick.net" no


In [22]:
# Proses Web Scraping Produk Seblak dari Tokopedia
from selenium import webdriver
from bs4 import BeautifulSoup
import pandas as pd
import time

data = []

driver = webdriver.Chrome()

for i in range(1,11):
    url = "https://www.tokopedia.com/search?navsource=&page={}&q=seblak&srp_component_id=02.01.00.00&srp_page_id=&srp_page_title=&st=".format(i)
    driver.get(url)
    html = driver.page_source
    soup = BeautifulSoup(html, "html.parser")
    time.sleep(2)
    
    for item in soup.findAll('div','css-uwyh54')[:10]:
        nama_produk = item.find('div','css-3um8ox').text
        harga_produk = item.find('div','css-1ksb19c').text

        for item1 in item.findAll('div','css-1rn0irl'):
            penjual = item1.findAll('span','css-1kdc32b')[1].text
            kota_toko = item1.findAll('span','css-1kdc32b')[0].text
        
        terjual= item.findAll('span','css-1duhs3e')
        if len(terjual)>0:
            banyak_terjual = item.find('span','css-1duhs3e').text
        else:
            banyak_terjual = None 

        rating=  item.findAll('span','css-t70v7i')
        if len(rating)>0:
            rating_produk = item.find('span','css-t70v7i').text
        else:
            rating_produk = None
    
        data.append((nama_produk, harga_produk, penjual, kota_toko, banyak_terjual, rating_produk))

driver.quit()

df = pd.DataFrame(data, columns = ['nama_produk', 'harga_produk', 'penjual', 'kota_toko', 'terjual', 'rating_produk'])

In [23]:
df

nama_produk harga_produk  \
0                    Mommy Seblak Basah Instant 100gr     Rp10.000   
1   Seblak Instan Ceu Nthien Khas Bandung Rasana N...     Rp15.000   
2                             SEBLAK INSTANT TERMURAH      Rp3.999   
3                          seblak instant sajodo snak     Rp19.000   
4   SEBLAK VIRAL - SEBLAK INSTANT - SEBLAK RAFAEL ...      Rp8.880   
..                                                ...          ...   
95  cuanki tahu pasung isi 50 murah batagor kering...     Rp12.000   
96  KERUPUK SEBLAK 250gr / KRUPUK SEBLAK DAUN JERU...     Rp12.500   
97  Kerupuk Orange / Kerupuk Seblak 250gr / Kerupu...      Rp7.500   
98           Seblak kuah instan , Seblak khas garut b      Rp5.000   
99                   Paket seblak sosis untuk 3 porsi     Rp29.000   

                           penjual          kota_toko         terjual  \
0                         emmoshop         Kab. Bogor   Terjual 2 rb+   
1         Central Seblak Nusantara  Tangerang Selatan   Terjual 1 rb+   
2   Pusat Cuankie & Cemilan Frozen      Jakarta Timur  Terjual 10 rb+   
3               hikmah berkah jaya      Jakarta Barat    Terjual 750+   
4                  FARADHIFY STORE             Bekasi    Terjual 100+   
..                             ...                ...             ...   
95                 CuankiCulametan         Kab. Garut    Terjual 100+   
96                   Dapur Sehat21       Kab. Bandung    Terjual 100+   
97                     BawangBagus          Tangerang       Terjual 5   
98        iwan resistans leather 2         Kab. Garut    Terjual 500+   
99               Tukangsayurmbakyu  Tangerang Selatan       Terjual 8   

   rating_produk  
0            4.9  
1            4.9  
2            4.8  
3            4.8  
4            4.8  
..           ...  
95           4.9  
96           4.9  
97          None  
98           4.9  
99           5.0  

[100 rows x 6 columns]

Dari proses web scraping yang sudah dilakukan, hasil yang didapatkan adalah 100 produk dari 10 halaman dimana perhalaman saya menetapkan limit sebanyak 10 produk.

# B. Data Preparation

In [24]:
# Mencari info dari dataframe yang telah dibuat dari hasil scraping.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   nama_produk    100 non-null    object
 1   harga_produk   100 non-null    object
 2   penjual        100 non-null    object
 3   kota_toko      100 non-null    object
 4   terjual        91 non-null     object
 5   rating_produk  88 non-null     object
dtypes: object(6)
memory usage: 4.8+ KB


In [26]:
# Melakukan cek terhadap duplikat.
df.drop_duplicates(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99 entries, 0 to 99
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   nama_produk    99 non-null     object
 1   harga_produk   99 non-null     object
 2   penjual        99 non-null     object
 3   kota_toko      99 non-null     object
 4   terjual        90 non-null     object
 5   rating_produk  87 non-null     object
dtypes: object(6)
memory usage: 5.4+ KB


Bisa dilihat jika jumlah entries yang tadinya 100 sekaarang berkurang menjadi 99 setelah 1 data yang duplicate dibuang.

In [27]:
# Handle missing values
df[['terjual','rating_produk']] = df[['terjual','rating_produk']].fillna(0)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99 entries, 0 to 99
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   nama_produk    99 non-null     object
 1   harga_produk   99 non-null     object
 2   penjual        99 non-null     object
 3   kota_toko      99 non-null     object
 4   terjual        99 non-null     object
 5   rating_produk  99 non-null     object
dtypes: object(6)
memory usage: 5.4+ KB


Setelah missing values diisi menjadi '0' bisa dilihat jika nilai null pada kolom terjual dan rating_produk menjadi hilang.

In [28]:
# Handle character
cols = ['harga_produk','terjual','rating_produk']
df[cols] = df[cols].replace({'\+':'','Terjual':'','Rp':'',' rb':'000'}, regex=True)
df

nama_produk harga_produk  \
0                    Mommy Seblak Basah Instant 100gr       10.000   
1   Seblak Instan Ceu Nthien Khas Bandung Rasana N...       15.000   
2                             SEBLAK INSTANT TERMURAH        3.999   
3                          seblak instant sajodo snak       19.000   
4   SEBLAK VIRAL - SEBLAK INSTANT - SEBLAK RAFAEL ...        8.880   
..                                                ...          ...   
95  cuanki tahu pasung isi 50 murah batagor kering...       12.000   
96  KERUPUK SEBLAK 250gr / KRUPUK SEBLAK DAUN JERU...       12.500   
97  Kerupuk Orange / Kerupuk Seblak 250gr / Kerupu...        7.500   
98           Seblak kuah instan , Seblak khas garut b        5.000   
99                   Paket seblak sosis untuk 3 porsi       29.000   

                           penjual          kota_toko terjual rating_produk  
0                         emmoshop         Kab. Bogor    2000           4.9  
1         Central Seblak Nusantara  Tangerang Selatan    1000           4.9  
2   Pusat Cuankie & Cemilan Frozen      Jakarta Timur   10000           4.8  
3               hikmah berkah jaya      Jakarta Barat     750           4.8  
4                  FARADHIFY STORE             Bekasi     100           4.8  
..                             ...                ...     ...           ...  
95                 CuankiCulametan         Kab. Garut     100           4.9  
96                   Dapur Sehat21       Kab. Bandung     100           4.9  
97                     BawangBagus          Tangerang       5             0  
98        iwan resistans leather 2         Kab. Garut     500           4.9  
99               Tukangsayurmbakyu  Tangerang Selatan       8           5.0  

[99 rows x 6 columns]

Proses ini dilakukan untuk menghilangkan character-character yang dapat membuat gangguaan saat memproses data seperti char '+','rb' dsb.

In [29]:
# Mengubah Dtype
df['terjual'] = df['terjual'].astype(int)
df['rating_produk'] = df['rating_produk'].astype(float)

In [30]:
# Mengubah Dtype
df['harga_produk'] = df['harga_produk'].str.replace('.', '').astype(int)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 99 entries, 0 to 99
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   nama_produk    99 non-null     object 
 1   harga_produk   99 non-null     int32  
 2   penjual        99 non-null     object 
 3   kota_toko      99 non-null     object 
 4   terjual        99 non-null     int32  
 5   rating_produk  99 non-null     float64
dtypes: float64(1), int32(2), object(3)
memory usage: 4.6+ KB


Bisa dilihat setelah proses penggantian Dtype, Dtype yang tadinya semuanya merupakan object sekarang sudah sesuai dengan semestinya.

# C. Business Understanding/Problem Statement

### SMART Framework

Case : Saya ingin menjalankan skema dropship menjual seblak namun masih belum yakin dan masih ingin mengetahui bagaimana penjualan seblak, apakah penjualannya tinggi, apakah banyak orang yang suka dsb.

- Specific : Saya ingin meningkatkan pendapatan dengan menjalankan bisnis dropshipping seblak.

- Measurable : Meningkatkan pendapatan sebesar 20%.

- Achievable : Dengan melakukan dropshipping, miningkatkan pendapatan sebesar 20% dapat dicapai.

- Relevant : Dengan melakukan dropshipping, hasil keuntungan yang didapat sudah jelas akan berpengaruh dengan tingkat pendapatan saya.

- Time-Bound : Target saya untuk menambah pendapatan dengan melakukan drop shipping dapat dilakukan dalam 6 bulan.

Saya ingin meningktakan pendapatan pribadi sebesar 20% dengan melakukan dropshipping produk seblak dalam rentan waktu 6 bulan. 



# D. Analysis

### 1. Mencari tahu nilai Mean, Median, Standar Deviasi, Skewness dan Kurtosis dari harga produk, rating produk dan banyak terjual produk seblak

In [31]:
# Harga Produk
print('mean dari data adalah =',df.harga_produk.mean())
print('median dari data adalah =',df.harga_produk.median())
print('standar deviasi dari data adalah =',df.harga_produk.std())
print("Skewness =", df.harga_produk.skew())
print("Kurtosis =", df.harga_produk.kurtosis())


mean dari data adalah = 22498.272727272728
median dari data adalah = 15000.0
standar deviasi dari data adalah = 22386.53645168032
Skewness = 2.349647081443584
Kurtosis = 5.593387188273972


Kesimpulan dari perhitungan data harga produk yang didapat adalah :
- Nilai std deviasi bisa dibilang baik karena nilai std deviasi lebih kecil dari  nilai mean.
- Tingkat skewness termasuk highly skewed karena nilai lebih dari 1.
- Nilai kurtosis masuk kedalam Leptokurtic karena nilai > 3.

In [32]:
# Terjual
print('mean dari data adalah =',df.terjual.mean())
print('median dari data adalah =',df.terjual.median())
print('standar deviasi dari data adalah =',df.terjual.std())
print("Skewness =", df.terjual.skew())
print("Kurtosis =", df.terjual.kurtosis())

mean dari data adalah = 426.5050505050505
median dari data adalah = 70.0
standar deviasi dari data adalah = 1462.1102541813991
Skewness = 5.9303355518438705
Kurtosis = 36.93851708986021


Kesimpulan dari perhitungan data terjual yang didapat adalah :
- Data yang didapat memiliki variasi yang tinggi dan lumayan tersebar karena nilai std deviasi lebih besar dari nilai mean.
- Tingkat skewness termasuk highly skewed karena nilai lebih dari 1.
- Nilai kurtosis masuk kedalam Leptokurtic karena nilai > 3.

In [33]:
# Rating Produk
print('mean dari data adalah =',df.rating_produk.mean())
print('median dari data adalah =',df.rating_produk.median())
print('standar deviasi dari data adalah =',df.rating_produk.std())
print("Skewness =", df.rating_produk.skew())
print("Kurtosis =", df.rating_produk.kurtosis())

mean dari data adalah = 4.312121212121212
median dari data adalah = 4.9
standar deviasi dari data adalah = 1.6125320896837188
Skewness = -2.3406167494140644
Kurtosis = 3.5831025050624046


Kesimpulan dari perhitungan data rating produk yang didapat adalah :
- Nilai std deviasi bisa dibilang baik karena nilai std deviasi lebih kecil dari  nilai mean.
- Tingkat skewness termasuk highly skewed karena nilai lebih dari -1.
- Nilai kurtosis masuk kedalam Mesokurtic karena nilai = 3.

### 2. Mencari tahu potensi pendapatan tertinggi hingga terendah dari penjualan produk seblak.

In [74]:
# Import Library
from scipy import stats
import numpy as np

# Mencari Upper dan Lower dari penjualan
std = df['terjual'].std()
N = len(df)
low, up = stats.norm.interval(0.95,loc=df['terjual'].mean(),scale=std/np.sqrt(N))
print('Lower Limit:',low)
print('Upper Limit:',up)

Lower Limit: 138.49302816680586
Upper Limit: 714.5170728432952


Bisa dilihat jika produk seblak akan terjual paling sedikit sebanyak 138 dan paling banyak akan mencapai 714 produk terjual.

In [84]:
# Hassil lower dan upper dikalikan dengan rata-rata harga produk
Minimal_Keuntungan = df.harga_produk.mean()*low
print('Perkiraan Minimal Keuntungan adalah =', Minimal_Keuntungan)
Maksimal_Keuntungan = df.harga_produk.mean()*up
print('Perkiraan Maksimal Keuntungan adalah =', Maksimal_Keuntungan)


Perkiraan Minimal Keuntungan adalah = 3115853.918522662
Perkiraan Maksimal Keuntungan adalah = 16075399.973121049


Setelah mengetahui berapa banyak produk yang bisa terjual maka hasil tersebut dikalikan dengan rata-rata harga produk yang menghasilkan perkiraan pendapatan minimal sebesar Rp3.115.853 dan perkiraan paling besar yaitu Rp16.075.399

### 3. Perbandingan harga penjualan produk seblak di Jabodetabek dan di luar Jabodetabek.

Hipotesis
- H0 = Harga Produk di Jabodetabek == Non-Jabodetabek.
- H1 = Harga Produk di Jabodetabek != Non-Jabodetabek.


Dari case tersebut maka saya akan melakukan test melakukan Two-Sample Independent Test karena ingin melakukan perbandingan terhadap 2 variabel yang tidak saling berkegantungan.

In [69]:
# Slicing kota yang termasuk di Jabodetabek
cities = ['Kab. Bogor', 'Tangerang Selatan', 'Jakarta Timur', 'Jakarta Barat', 'Bekasi',
          'Jakarta Selatan', 'Kab. Bekasi', 'Kab. Tangerang', 'Jakarta Utara', 'Jakarta Pusat', 'Tangerang', 'Depok']
Jabodetabek = df[df['kota_toko'].isin(cities)]

Jabodetabek.head()

nama_produk  harga_produk  \
0                   Mommy Seblak Basah Instant 100gr         10000   
1  Seblak Instan Ceu Nthien Khas Bandung Rasana N...         15000   
2                            SEBLAK INSTANT TERMURAH          3999   
3                         seblak instant sajodo snak         19000   
4  SEBLAK VIRAL - SEBLAK INSTANT - SEBLAK RAFAEL ...          8880   

                          penjual          kota_toko  terjual  rating_produk  
0                        emmoshop         Kab. Bogor     2000            4.9  
1        Central Seblak Nusantara  Tangerang Selatan     1000            4.9  
2  Pusat Cuankie & Cemilan Frozen      Jakarta Timur    10000            4.8  
3              hikmah berkah jaya      Jakarta Barat      750            4.8  
4                 FARADHIFY STORE             Bekasi      100            4.8

In [71]:
# Slicing kota yang tidak termasuk di Jabodetabek
Luar_Jabodetabek = df[~df['kota_toko'].isin(cities)]

Luar_Jabodetabek.head()

nama_produk  harga_produk  \
5   SEBRING KRUPUK KERUPUK SEBLAK KERING PEDAS DAU...         16000   
8   GURILEM MINI 1KG Siomay Kering Toping Baso Aci...         29999   
10                                     Kerupuk Seblak         20000   
12                                   seblak Noni noni         41500   
13  SUKRO CIKUR RENYAH NGEPROS isi 1KG Pelengkap B...         31999   

            penjual        kota_toko  terjual  rating_produk  
5       Aydaa Snack        Surakarta       40            4.9  
8       Pedas Gilaa         Surabaya      500            4.9  
10     Jajanan Wawa  Kab. Purwakarta        0            0.0  
12  Basrengsikembar     Kab. Bandung        2            0.0  
13        gaiagarut       Kab. Garut       60            5.0

In [68]:
# Mencari T-Stat dan P-Value
t_stat, p_val = stats.ttest_ind(Jabodetabek['harga_produk'], Luar_Jabodetabek['harga_produk'])

print('T-Statistic:', t_stat)
print('P-value:', p_val)

T-Statistic: -0.15818680663460305
P-value: 0.8746384524490391


Dari hasil perhitungan yang didapat dapat disimpulkan bahwa, karena P-value lebih dari 0.05 maka H0 diterima yang berarti harga produk di jabodetabek dan non_jabodetabek tidak memiliki perbedaan yang signifikan.

### 4. Mencari tahu apakah harga seblak yang murah berpengaruh terhadap keputusan seseorang untuk membeli seblak.

Hipotesis
- H0 = Tidak ada korelasi antara harga dengan jumlah pembeli.
- H1 = Ada korelasi anatara harga dengan jumlah pembeli.

Dalam case ini saya melakukan test korelasi menggunakan teknik Pearson karena kedua variabel bisa dibilang linear dan memiliki tingkat distribusi yang normal.

In [75]:
corr_r, pval_p = stats.pearsonr(df['harga_produk'], df['terjual'])

print(f"r-correlation: {corr_r:.2f}, p-value: {pval_p}")

r-correlation: -0.07, p-value: 0.47958728775050696


Dari hasil yang sudah didapatkan menunjukan nilai r-correlation -0.07 maka harga yang murah tidak berpengaruh signifikan dalam tingkat pembelian seblak dan nilai p-value sebesar 0.47 dimana hasil tersebut tidak cukup kuat untuk menolak hipotesis H0.

# E. Conclusion

Dari keseluruhan hasil yang didapat dapat diambil kesimpulan bahwa:
- Harga jual produk seblak memiliki harga yang bervariatif namun tetap pada range harga tertentu.
- Keseluruhan penjualan produk seblak cukup tinggi namun memiliki persebaran tingkat penjualan yang sangat bervariatif mulai dari sangat banyak, banyak hingga sedikit yang terjual atau bahkan belum terjual sama sekali.
- Kebanyakan orang menyukai produk seblak karena secara keseluruhan rata-rata penilaian orang terhadap seblak cukup baik.
- Potensi pendapatan maksimal dan minimal dari penjualan produk seblak mempunyai perbedaan yang signifikan begitupun dengan tingkat low dan up penjualannya. Namun secara keseluruhan jumlah penjualan produk seblak bisa dibilang cukup tinggi.
- Harga jual produk seblak cenderung sama dan tidak memiliki perbedaan yang sginifikan walaupun dijual dari kota yang berada di luar pulau jawa sekalipun.
- Harga jual seblak yang murah tidak memiliki pengaruh signifikan terhadap tingkat pembelian seblak, itu menunjukan memang produk seblak banyak digemari oleh orang-orang khusunya di Indonesia.


Dari beberapa kesimpulan tersebut jika kita kembali pada case SMART diatas, dengan menjalankan dropshiping seblak merupakan langkah yang pas jika ingin menambah uang saku karena daya beli produk seblak yang cukup besar, jangkauan pembeli yang luas dan harga yang bervariatif.